src/combobox/ComboBox.ts

```diff
@@ -0,0 +1,315 @@
+import { WidgetBase } from '@dojo/widget-core/WidgetBase';
+import { ThemeableMixin, ThemeableProperties, theme } from '@dojo/widget-core/mixins/Themeable';
+import { FormLabelMixinProperties } from '@dojo/widget-core/mixins/FormLabel';
+import { v } from '@dojo/widget-core/d';
+import { DNode, HNode } from '@dojo/widget-core/interfaces';
+import { assign } from '@dojo/core/lang';
+
+import * as css from './styles/comboBox.css';
+
+/**
+ * @type ComboBoxProperties
+ *
+ * Properties that can be set on a ComboBox component
+ *
+ * @property autoBlur			Determines whether the input should blur after value selection
+ * @property clearable			Determines whether the input should be able to be cleared
+ * @property inputProperties	HTML properties supported by FormLabelMixin to set on the underlying input
+ * @property openOnFocus		Determines whether the result list should open when the input is focused
+ * @property results			Results for the current search term; should be set in response to `onRequestResults`
+ * @property value				Value to set on the input
+ * @property getResultValue		Can be used to get the text value of a result based on the underlying result object
+ * @property onBlur				Called when the input is blurred
+ * @property onChange			Called when the value changes
+ * @property onFocus			Called when the input is focused
+ * @property onRequestResults	Called when results are shown; should be used to set `results`
+ * @property onMenuChange		Called when menu visibility changes
+ * @property renderMenu			Can be used to render a custom result menu
+ * @property renderResult		Can be used to render a custom result
+ */
+export interface ComboBoxProperties extends ThemeableProperties {
+	autoBlur?: boolean;
+	clearable?: boolean;
+	inputProperties?: FormLabelMixinProperties;
+	openOnFocus?: boolean;
+	results?: any[];
+	value?: string;
+	getResultValue?(result: any): string;
+	onBlur?(value: string): void;
+	onChange?(value: string): void;
+	onFocus?(value: string): void;
+	onRequestResults?(value: string): void;
+	onMenuChange?(open: boolean): void;
+	renderMenu?(resultItems: any[]): DNode;
```

Comment:
Did we think about other approaches that could be used to enable consumers to provide custom widgets? If the default Menu and Item were converted to widgets we could use the scoped registry (provided by the RegistryMixin) to swap the menu and/or item for a custom widget provided via the `ComboBoxProperties`. That feels pretty clean to me and would also promote reuse of the custom widgets more so than passing a couple of custom functions?